widget/textgrid.go

```diff
@@ -15,6 +15,7 @@ const (
 	textAreaSpaceSymbol   = '·'
 	textAreaTabSymbol     = '→'
 	textAreaNewLineSymbol = '↵'
+	textTabIndent         = "    "
```

Comment:
I don't think this is needed

Reply:
Hmm, I agree that the constant should be replaced by a variable based on the TabWidth setting, but the whole tab in a textgrid has to to be looked at. 
In `SetText()` every '\t' is replaced by 4 cells of ' '
Then in `textGridRenderer.refreshGrid()` if ShowWhitespace is true and the cell is '\t' it is replaced with the textAreaTabSymbol which can never happen as there are no '\t' in the cells.

To make the textGrid handle tabs the same as the rendered text will take a bit more work. It would have to handle the "12\t567" scenario and making tabs visible would be interesting!? I think it would involve inverting the white space handling. Ie replace all whitespace with the visible characters, which would expand the '\t' to the relevant number of textAreaTabSymbol. Then in refreshGrid() convert them back to whitespace if ShowWhiteSpace=false.

Assuming of course you still want visible whitespace and tab handling in textGrid?

Reply:
> if ShowWhitespace is true and the cell is '\t' it is replaced with the textAreaTabSymbol which can never happen as there are no '\t' in the cells.

That should not be the case - we compare a `\t` in the source `string`, it is not subject to the same string rendering that you have updated for the `canvas.Text`.

TextGrid does not render a literal `\t` character - it puts the `\t` in the `Rune` field, but shows either a space or a symbol based on the whitespace mode.
It leaves a certain number of characters to emulate the same indentation as you have coded into Text.
To complete it here the number it moves should be %numberOfSpaces.

The reason I commented on this line addition is that I thought it was not needed.
Either we leave this in and make another PR to update TextGrid calculations, or we can get this fixed up here...

Reply:
The `TextGrid` `SetText` method could be updated without this `textTabIntent`, something like this:
```go
func (t *TextGrid) SetText(text string) {
	lines := strings.Split(text, "\n")
	rows := make([]TextGridRow, len(lines))
	for i, line := range lines {
		lineLen := len(line)
		for _, r := range line {
			if r == '\t' {
				len += tabSpaceCount-1
			}
		}

		cells := make([]TextGridCell, lineLen)
		extras := 0
		for j, r := range line {
			cells[j+extras] = TextGridCell{Rune: r}

			if r == '\t' {
				for k := j + extras + 1; k < j+extras+tabSpaceCount; k++ {
					cells[k] = TextGridCell{Rune: ' '}
				}
				extras += tabSpaceCount - 1
			}
		}

		rows[i] = TextGridRow{Cells: cells}
	}

	t.Rows = rows
	t.Refresh()
}
```

Reply:
I have implemented tab stops in TextGrid. Tab characters are now padded with spaces to the next tab stop instead of being padded with 3 spaces. This now matches the behaviour of the canvas.Text().
One thing I am unsure about is how I have set the tabWidth. I have added it to `TextGrid` and not to the `TextGridStyle` this is because of how TextGrid works.  When `SetText()` is called the tabs are padded with spaces and rune by rune written to the grid cells. Conversely when `Text()` is called the runes are read from the grid cell by cell and tab padding spaces are removed. This means tab padding spaces are 'hard coded' into the cells it is not possible to update the tabWidth after the text has been set. As the style can be changed at any time it is not a suitable place to set the tabWidth. 
I did look at the possibility of updating the text with every style change but this gets complicated because of the ability to change the cell contents without using `SetText()`. i.e it is possible to insert unpadded tab characters or replace padding spaces with other characters.

Reply:
Hmm, this is tricky.
How about caching a copy of the current tab width value - then we can see if it changed on refresh.
If it did then we could iterate the rows of cells. If we use the tab indicator rune to show each tab location it could be re-arranged in the line I think? (possibly a lot more work?)

Reply:
I thought about using a special rune perhaps the HT U+2409 but no matter what symbol you use you end up with the problem "what if someone wants to display symbol X"
I considered keeping a text version of the rows and refreshing them on each `SetTabWidth()` but then you have the problem of two sources of truth. Which is the master value, the text line or cell value?
To do it properly really means separating the data and display. Then only edit the data and refresh the display on each update. Which is really "let's write an editor".
This TextGrid is a strange beast, neither fish nor fowl. Is it looks like a spreadsheet but is used as a text field. If it was a spreadsheet it would not need to support tabs, if it was a text field it would not need to support cell updates.

Reply:
I think this text const could be removed now?